Lua debugger server

aryla opened the pull request:
I don't know if this is something you want as part of ITGmania, but I wrote a Lua debugger server. This has been very useful for debugging theme issues.

## Demo video

[Shows off some of the debugger features, such as setting breakpoints, stopping at a breakpoint, stepping through the code, and inspecting variables.](https://github.com/user-attachments/assets/3942fd17-12e6-435b-aca0-4c3ae730a581)

## Configuration

The server implements the [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/), so any IDE that implements it should work as a frontend. I've tested this on Linux and Windows with VSCode and Zed.

Extension for VSCode: [vscode-lua-debugger-for-itgmania](https://github.com/aryla/vscode-lua-debugger-for-itgmania/releases/tag/v0.0.1).

Extension for Zed: [zed-lua-debugger-for-itgmania](https://github.com/aryla/zed-lua-debugger-for-itgmania/releases/tag/v0.0.1).

Example configuration for Neovim (with nvim-dap plugin):
```lua
dap.adapters.itgmania_lua = {
    type = 'server',
    host = '127.0.0.1',
    port = 8173,
}

dap.configurations.lua = {
    {
        type = 'itgmania_lua',
        request = 'attach',
        name = 'Debug ITGmania Lua code',
    }
}
```

## Features
- Pausing and stepping
- Breakpoints
- Conditional breakpoints
- Log points
- Expression evaluation
- Viewing and setting variables

## New command line options
- `--lua-debugger[=host:port]`
    Enable the debugger server. Defaults to "localhost:8173" if the host and port are not specified.

- `--lua-debugger-paused`
    Pause at first Lua code execution.

## Known issues
- Slows down Lua execution if there are any breakpoints set. There are some optimizations that could be done to improve the performance.
- When setting a breakpoint, there's no feedback on if it's on a valid line or not.
- Gameplay is not paused when stopped on a breakpoint so this is not that useful for debugging song Lua code or gameplay-related theme code.
- I haven't tested this on Mac OS, but at least it compiles in the pipeline.

## Typical sequence of events

1. `sm_main`, on main thread
The program sees the `--lua-debugger` command line flag and launches LuaDebugManager.
1. `LuaDebugManager::Start`, on main thread
LuaDebugManager starts the server, which listens for connections on a separate thread.
1. `LuaManager::Get`, on Lua thread [^1]
Meanwhile, LuaManager helps LuaDebug::Debugger keep track of all lua\_State objects by calling LuaDebugManager::ActivateState whenever a thread wants to run Lua code.
1. `LuaDebug::Server::handleConnection`, on connection handler thread
When a client connects, the server spawns a new thread to handle the connection and calls handleConnection. LuaDebug::Server passes requests from the client to LuaDebug::Debugger, and responses and events from Debugger to the client.
1. `LuaDebug::Debugger::SendRequest`, on connection handler thread
Debugger decides whether the request should be handled on the connection handler thread or on the Lua thread. If Lua execution is paused, it handles the request on the Lua thread, otherwise on the connection handler thread.
1. `LuaDebug::Debugger::HandleRequest`, on connection handler thread or Lua thread
Debugger dispatches the request to one of the handler functions.
1. `LuaDebug::Debugger::UpdateHook`, on connection handler thread or Lua thread
When a breakpoint is added, Debugger calls lua\_sethook to set the debug hook that is called for each line of Lua code. [^2] [^3] [^4]
1. `LuaDebug::Debugger::LuaHook`, on Lua thread
When the debug hook finds that a breakpoint is hit, it sends a `stopped` event and starts handling requests until execution is continued again.

[^1]: Lua code can be executed on different threads, but because of the lock in LuaManager, only one can run at a time. Here "Lua thread" refers to whatever thread is currently running Lua code.
[^2]: The event mask is set depending on what breakpoints there are. The hook can be called at each line, each return or each function call.
[^3]: This also applies to commands like `pause` and `stepOut` since they are implemented using breakpoints.
[^4]: This is a little bit sketchy since there is no locking in lua\_sethook and the lua\_State might be executing Lua code, but I don't see any other way to do it. So far I haven't ran into any issues because of this.

## Code overview

**LuaDebugManager.h/.cpp**: Existing code (sm\_main and LuaManager) interacts with the Lua debugger through the LuaDebugManager class. The LuaDebug::Server class handles communication with the client. 

**LuaDebugger.h/.cpp**: The LuaDebug::Debugger class executes commands received from the client and contains the Lua debug hook.

**LuaDebuggeeState.h/.cpp**: The LuaDebug::DebuggeeState class stores the state of the Lua execution environment (things like stack frames, locals, globals) while it's being inspected in the debugger. The state is only available when execution is paused.

**LuaDebugBreakpoint.h/.cpp**: The LuaDebug::Breakpoint class represents a breakpoint in Lua code. It handles evaluation of breakpoint conditions.

**LuaDebugDap.h/.cpp**: Debug Adapter Protocol data structures. LuaDebug::Header is the message header. Request, Response, and Event are kinds of messages that can be sent or received. They are actually just Json::Value objects.

**LuaDebugHelpers.h/.cpp**: Helper functions.

## v2

- Changed to use IXWebSocket.
- Fixed a missing "expensive" property in the response for the "scopes" commmand.
- Fixed the server sending events before the "initialize" request in some cases.
- The server now always tries to send a "terminated" event before closing the connection. Zed has some problems when a connection is closed without the event.

## v3

- Rebased and fixed conflicts.
- Run clang-format.

teejusb commented:
Since this looks to be totally optional (controlled by command line flags), seems fine to me. I'm not sure how much ill *personally* use it but maybe some people might want it and seems pretty out of the way from the critical path.

One question, do we need to add another socket implementation or can we reuse the network manager somehow (which has websocket support already). I'd prefer to consolidate.

aryla commented:
Great!
> One question, do we need to add another socket implementation or can we reuse the network manager somehow

Good point, looks like the IXWebSocket library that the NetworkManager uses could work for this too. I'll change this to use that for the sockets. Thanks for the suggestion!

aryla committed:
LuaDebug: Add a debugger server for Lua

Adds a Lua debugger server. The server implements the Debug Adapter
Protocol so any IDE or editor can be used as the frontend.

Features:
    - Pausing and stepping
    - Breakpoints
    - Conditional breakpoints
    - Log points
    - Expression evaluation
    - Viewing and setting variables

New command line options:
    --lua-debugger[=host:port]
        Enable the debugger server. Defaults to "localhost:8173" if the
        host and port are not specified.

    --lua-debugger-paused
        Pause at first Lua code execution.

teejusb commented:
Didn't do a supperrr deep review but doing a short read through does look reasonable to me! Since it's not on the critical path and totally optional I'll go ahead and merge!

teejusb commented:
I think it would be super useful to add documentation in the ITGmania repo itself too for others in the future :)

aryla commented:
Thanks! I'll write some documentation.
